src/resources/hash.cpp

```diff
@@ -0,0 +1,163 @@
+#include "resources/hash.hpp"
+#include "fileutils.hpp"
+#include "sha256.h"
+#include <memory>
+#include <optional>
+
+using namespace buddy::resources;
+
+class HashContext {
+private:
+    mbedtls_sha256_context sha;
+    uint32_t counter;
+
+public:
+    HashContext()
+        : sha()
+        , counter(0) {
+        mbedtls_sha256_init(&sha);
+        mbedtls_sha256_starts_ret(&sha, 0);
+    }
+
+    void append_mark() {
+        mbedtls_sha256_update_ret(&sha, (unsigned char *)&counter, sizeof(counter));
+        counter += 1;
+    }
+
+    void append_data(const uint8_t *data, size_t length) {
+        mbedtls_sha256_update_ret(&sha, data, length);
+    }
+
+    void finish(Hash &output) {
+        mbedtls_sha256_finish_ret(&sha, &output[0]);
+    }
+
+    ~HashContext() {
+        mbedtls_sha256_free(&sha);
+    }
+};
+
+static bool update_hash_with_file(const char *root, Path &path, HashContext &hash_ctx) {
+    // mark
+    hash_ctx.append_mark();
+
+    // filepath
+    size_t root_length = strlen(root);
+    const char *full_path = path.get();
+    const char *filepath = full_path + root_length;
+    hash_ctx.append_data((uint8_t *)filepath, strlen(filepath));
+
+    // mark
+    hash_ctx.append_mark();
+
+    // filedata
+    std::unique_ptr<FILE, FILEDeleter> source(fopen(path.get(), "rb"));
+    if (source.get() == nullptr) {
+        log(LOG_SEVERITY_ERROR, "Failed to open file %s", source.get());
```

Comment:
Could we extract this `log(...)` to a common place? Most of the failed `fopen` scenarios need logging of the error, at least the filename. And it would be nice to have all the error texts consistent.

One possible solution would be using some custom class with a bool CheckSuccessfullyOpenned() instead of a std::unique_ptr.

Reply:
You mean to wrap `FILE` with our custom wrapper class?

I am not against! It sounds to me as something that needs to be carefully designed and then be available for the whole codebase. I don't really want to introduce classes like this just for the sake of the _resources_ module unless necessary. 

However, when implementing such a class, I am not convinced the right approach is to log all its unsuccessful `open()` calls. First, is unsuccessful call to `open` always a failure to be reported? I don't think so. For example, you might be using it to check whether a file exists. Second, having consistent logs from such module (and not logging it as the caller) might be a downside. Yes, it is a less code. But the logs from the caller (not the callee) give you much more information what went wrong. And the slight differences also enable you trace where exactly it went wrong. 

I am not trying to say this code is the best. I just don't think introducing a whole new class for such problem is necessary better. Maybe just a macro to report the error would be enough in this case? 🤷‍♂️ We can discuss in person ☺️

Reply:
ok